File: runtime/dart/convert/closableStringSink.ts
import { IIterable } from "../core/iterable";
import { IStringSink } from "../core/stringSink";
declare const dart: {
    convert: {
        closableStringSinkFromStringSink: (
            sink: IStringSink,
            onClose: () => void
        ) => IClosableStringSink;
    };
};
export interface IClosableStringSink {
    close: () => void;
    write: (object?: Object | undefined) => void;
    writeAll: (objects: IIterable<any>, separator: string) => void;
    writeln: (object?: Object | undefined) => void;
    writeCharCode: (charCode: number) => void;
    toString: () => string;
    getHashCode: () => number;
}
export class ClosableStringSink implements IStringSink {
    public static fromStringSink(
        sink: IStringSink,
        onClose: () => void
    ): IClosableStringSink {
        return dart.convert.closableStringSinkFromStringSink(sink, onClose);
    }
    private readonly _dart_close: () => void = undefined as any;
    private readonly _dart_write: (object?: Object | undefined) => void =
        undefined as any;
    private readonly _dart_writeAll: (
        objects: IIterable<any>,
        separator: string
    ) => void = undefined as any;
    private readonly _dart_writeln: (object?: Object | undefined) => void =
        undefined as any;
    private readonly _dart_writeCharCode: (charCode: number) => void =
        undefined as any;
    private readonly _dart_toString: () => string = undefined as any;
    private readonly _dart_getHashCode: () => number = undefined as any;
    public close(): void {
        return this._dart_close();
    }
    public write(object?: Object | undefined): void {
        return this._dart_write(object);
    }
    public writeAll(objects: IIterable<any>, separator: string = ""): void {
        return this._dart_writeAll(objects, separator);
    }
    public writeln(object?: Object | undefined = ""): void {
        return this._dart_writeln(object);
    }
    public writeCharCode(charCode: number): void {
        return this._dart_writeCharCode(charCode);
    }
    public toString(): string {
        return this._dart_toString();
    }
    public getHashCode(): number {
        return this._dart_getHashCode();
    }
}

File: lib/cfr/builtins/libs/dart/convert/closableStringSink.dart
import 'dart:convert';
import 'dart:core';

import 'package:hydro_sdk/cfr/runtimeSupport.dart';

class VMManagedClosableStringSink extends VMManagedBox<ClosableStringSink> {
  VMManagedClosableStringSink(
      {required this.table, required this.vmObject, required this.hydroState})
      : super(
          table: table,
          vmObject: vmObject,
          hydroState: hydroState,
        ) {
    table['close'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      vmObject.close();
      return [];
    });
    table['write'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      vmObject.write(maybeUnBoxAndBuildArgument<Object?>(luaCallerArguments[1],
          parentState: hydroState));
      return [];
    });
    table['writeAll'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      vmObject.writeAll(
          maybeUnBoxAndBuildArgument<Iterable<dynamic>>(luaCallerArguments[1],
              parentState: hydroState),
          luaCallerArguments[2]);
      return [];
    });
    table['writeln'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      vmObject.writeln(maybeUnBoxAndBuildArgument<Object?>(
          luaCallerArguments[1],
          parentState: hydroState));
      return [];
    });
    table['writeCharCode'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      vmObject.writeCharCode(luaCallerArguments[1]);
      return [];
    });
    table['toString'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.toString(),
      ];
    });
    table['getHashCode'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.hashCode,
      ];
    });
  }

  final HydroTable table;

  final HydroState hydroState;

  final ClosableStringSink vmObject;
}

void loadClosableStringSink(
    {required HydroState hydroState, required HydroTable table}) {
  table['closableStringSinkFromStringSink'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    Closure unpackedonClose = luaCallerArguments[2];
    return [
      maybeBoxObject<ClosableStringSink>(
          object: ClosableStringSink.fromStringSink(
              maybeUnBoxAndBuildArgument<StringSink>(luaCallerArguments[1],
                  parentState: hydroState),
              () => unpackedonClose.dispatch(
                    [
                      luaCallerArguments[0],
                    ],
                    parentState: hydroState,
                  )),
          hydroState: hydroState,
          table: HydroTable()),
    ];
  });
  registerBoxer<ClosableStringSink>(boxer: (
      {required ClosableStringSink vmObject,
      required HydroState hydroState,
      required HydroTable table}) {
    return VMManagedClosableStringSink(
        vmObject: vmObject, hydroState: hydroState, table: table);
  });
}
